### PR TITLE
planner: fix stack overflow caused by folding constant

### LINKED ISSUE #10156: Stack overflow caused by checking constant wrongly
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```
CREATE TABLE `t1` (
  `period_name` varchar(24) DEFAULT NULL ,
  `period_id` bigint(20) DEFAULT NULL ,
  `starttime` bigint(20) DEFAULT NULL 
);

CREATE TABLE `t2` (
  `bussid` bigint(20) DEFAULT NULL,
  `ct` bigint(20) DEFAULT NULL
);

select
    a.period_name,
    b.date8
from
    (select * from t1) a
left join
    (select bussid,date(from_unixtime(ct)) date8 from t2) b
on 
    a.period_id = b.bussid
where 
    datediff(b.date8, date(from_unixtime(a.starttime))) >= 0;
```

2. What did you expect to see?
```
mysql> select
    ->     a.period_name,
    ->     b.date8
    -> from
    ->     (select * from t1) a
    -> left join
    ->     (select bussid,date(from_unixtime(ct)) date8 from t2) b
    -> on 
    ->     a.period_id = b.bussid
    -> where 
    ->     datediff(b.date8, date(from_unixtime(a.starttime))) >= 0;
Empty set (0.01 sec)
```

3. What did you see instead?
No result and eventually panic caused by stack overflow.
```
mysql> select     a.period_name,     b.date8 from     (select * from t1) a left join     (select bussid,date(from_unixtime(ct)) date8 from t2) b on      a.period_id = b.bussid where      datediff(b.date8, date(from_unixtime(a.starttime))) >= 0;
ERROR 2013 (HY000): Lost connection to MySQL server during query
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
```
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                                         |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v3.0.0-beta.1-109-g7666f688b
Git Commit Hash: 7666f688b3d91ea84fa6b0a20c2e784fff3c8945
Git Branch: master
UTC Build Time: 2019-04-15 12:14:36
GoVersion: go version go1.11.3 darwin/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve? 
Fix #10156.

### What is changed and how it works?
When creating dummy function in `foldConstant`, use `NewFunctionBase` to avoid it fold it again.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: sessionctx/stmtctx/stmtctx.go [hunk 1]
```diff
@@ -423,3 +426,45 @@ func (sc *StatementContext) ShouldIgnoreOverflowError() bool {
 	}
 	return false
 }
+
+// CopTasksDetails returns some useful information of cop-tasks during execution.
```
Comment (shenli) [inline sessionctx/stmtctx/stmtctx.go]: Is this related with this PR?

Comment (qw4990) [reply sessionctx/stmtctx/stmtctx.go]: My fault, it's a mistake.
I will fix it and cherry-pick to 2.1 after merging.

Comment (qw4990) [reply sessionctx/stmtctx/stmtctx.go]: Fixed.

**Comment by qw4990** [comment]
> Why in #10156 , you said we can use `ConstItem` to fix it, but here just replace `NewFunctionInternal` by `NewFunctionBase`?

Because `CorrelatedColumn.ConstItem()` always returns true now, it's not correct.
In fact, we have to differentiate between planner-stage and execution-stage for it.

In planner-stage, for example when building scalar function, it should return false because its result depend on column data.
In execution-stage, for example when building range in some operators, since the outer row has been chosen, we can regard it as a const.

I will fix this problem in another PR.
So for this problem, I fix it by the way for simplicity.

**Comment by qw4990** [comment]
PTAL~

**Comment by zz-jason** [comment]
This bug is introduced in https://github.com/pingcap/tidb/pull/7696
@qw4990 Please cherry pick this fix to the release 2.1 branch

**Pull request merged**